src/rdsn/src/utils/metrics.cpp

```diff
@@ -132,27 +185,78 @@ uint64_t percentile_timer::generate_initial_delay_ms(uint64_t interval_ms)
       _timer(new boost::asio::deadline_timer(tools::shared_io_service::instance().ios))
 {
     _timer->expires_from_now(boost::posix_time::milliseconds(_initial_delay_ms));
     _timer->async_wait(std::bind(&percentile_timer::on_timer, this, std::placeholders::_1));
 }
 
+void percentile_timer::close()
+{
+    // If the timer has already expired when cancel() is called, then the handlers for asynchronous
+    // wait operations will:
+    // * have already been invoked; or
+    // * have been queued for invocation in the near future.
+    //
+    // These handlers can no longer be cancelled, and therefore are passed an error code that
+    // indicates the successful completion of the wait operation. Thus set the state of timer to
+    // kClosing to tell on_timer() that the timer should be closed even if it is not called with
+    // operation_canceled.
+    auto expected_state = state::kRunning;
+    if (_state.compare_exchange_strong(expected_state, state::kClosing)) {
+        _timer->cancel();
+    }
+}
+
+void percentile_timer::wait() { _completed.wait(); }
+
+void percentile_timer::on_close()
+{
+    _on_close();
+    _completed.notify();
+}
+
 void percentile_timer::on_timer(const boost::system::error_code &ec)
 {
+// This macro is defined for the case that handlers for asynchronous wait operations are no
+// longer cancelled. It just checks the internal state atomically (since close() can also be
+// called simultaneously) for kClosing; once it's matched, it will not execute future callbacks
+// to stop the timer.
+#define TRY_PROCESS_TIMER_CLOSING()                                                                \
+    do {                                                                                           \
+        auto expected_state = state::kClosing;                                                     \
+        if (_state.compare_exchange_strong(expected_state, state::kClosed)) {                      \
+            on_close();                                                                            \
+            return;                                                                                \
+        }                                                                                          \
+    } while (0)
+
     if (dsn_unlikely(!!ec)) {
         dassert_f(ec == boost::system::errc::operation_canceled,
                   "failed to exec on_timer with an error that cannot be handled: {}",
                   ec.message());
+
+        // Cancel can only be launched by close().
+        auto expected_state = state::kClosing;
+        dassert_f(_state.compare_exchange_strong(expected_state, state::kClosed),
+                  "wrong state for percentile_timer: {}, while expecting closing state",
+                  static_cast<int>(expected_state));
+        on_close();
+
         return;
     }
 
-    _exec();
+    TRY_PROCESS_TIMER_CLOSING();
+    _on_exec();
 
+    TRY_PROCESS_TIMER_CLOSING();
```

Comment:
better to add `#undef TRY_PROCESS_TIMER_CLOSING` when you'll never use it.

Reply:
Sure.